### PR TITLE
Start of work for footnotes and abbr

### PR DESCRIPTION
**Pull request by artragis**
I commit some work for footnotes and abbr.

As I was coding I clearly saw that my approach for abbr is not the good one (no ctx))

Please @vhf can you give me some tricks to succeed?

### CODE REVIEW

File: packages/rebber/src/pre-visitors/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  'text': [abbrPlugin().txtVisitor],
+}
+
+module.exports = preVisit
+
+function preVisit (ctx, root) {
+  const visitors = xtend(defaultVisitors, ctx.preprocessors || {})
+  Object.keys(visitors).forEach((key) => {
+    // my old js says me it's bad but i have to try
+    if (typeof visitors[key] === typeof []) {
```
Comment (vhf) [inline packages/rebber/src/pre-visitors/index.js]: `if (Array.isArray(visitors[key])` 😉

File: packages/rebber/src/pre-visitors/index.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+const xtend = require('xtend')
+const visit = require('unist-util-visit')
+const abbrPlugin = require('abbr')()
+
+const defaultVisitors = {
+  'linkReference': [abbrPlugin().abbrVisitor],
+  'text': [abbrPlugin().txtVisitor],
```
Comment (vhf) [inline packages/rebber/src/pre-visitors/index.js]: 💯

File: packages/rebber/src/pre-visitors/index.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const xtend = require('xtend')
+const visit = require('unist-util-visit')
+const abbrPlugin = require('abbr')()
```
Comment (vhf) [inline packages/rebber/src/pre-visitors/index.js]: `const abbrPlugin = require('./abbr')()`

File: packages/rebber/src/types/linkReference.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const all = require('all')
```
Comment (vhf) [inline packages/rebber/src/types/linkReference.js]: `const all = require('./all')`

File: packages/rebber/src/pre-visitors/abbr.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+module.exports = plugin
+const identifiers = {}
+function plugin (ctx) {
+  return {
+    abbrVisitor: abbrVisitor,
+    txtVisitor: txtVisitor
+  }
+  function abbrVisitor (node, index, parent) {
+    if (node.referenceType === 'shortcut') { // remark for abbr
+      identifiers[node.identifier] = all(ctx || {}, node)
```
Comment (vhf) [inline packages/rebber/src/pre-visitors/abbr.js]: It'd be better to have `all(ctx, node` here and have a default parameter in the function definition: `function plugin (ctx = {}) {`

File: packages/rebber/src/index.js [hunk 6]
```diff
@@ -1,13 +1,14 @@
 /* Dependencies. */
 const xtend = require('xtend')
 const one = require('./one')
+const preprocess = require('pre-visitors')
```
Comment (vhf) [inline packages/rebber/src/index.js]: `const preprocess = require('./pre-visitors')`

File: packages/rebber/src/pre-visitors/index.js (hunk 1 above)
Comment (artragis) [inline packages/rebber/src/pre-visitors/index.js]: Thx

File: packages/rebber/src/types/linkReference.js (hunk 4 above)
Comment (artragis) [inline packages/rebber/src/types/linkReference.js]: yes

File: packages/rebber/src/index.js (hunk 6 above)
Comment (artragis) [inline packages/rebber/src/index.js]: ok

**Pull request merged**